src/passes/Outlining.cpp

```diff
@@ -13,23 +13,225 @@
+    if (builder) {
+      ASSERT_ERR(builder->visit(curr));
+    }
+    DBG(printVisitExpression(curr));
+
+    if (state == InSeq || state == InSkipSeq) {
+      maybeEndSeq();
+    }
+  }
+  // Helpers
+  void startExistingFunction(Function* func) {
+    ASSERT_ERR(existingBuilder.build());
+    ASSERT_ERR(existingBuilder.visitFunctionStart(func));
+    instrCounter = 0;
+    seqCounter = 0;
+    state = NotInSeq;
+    DBG(std::cout << "\n\n$" << func->name << " Func Start "
+                  << &existingBuilder);
+  }
+  ReconstructState getCurrState() {
+    // We are either in a sequence or not in a sequence. If we are in a sequence
+    // and have already created the body of the outlined function that will be
+    // called, then we will skip instructions, otherwise we add the instructions
+    // to the outlined function. If we are not in a sequence, then the
+    // instructions are sent to the existing function.
+    if (seqCounter < sequences.size() &&
+        instrCounter >= sequences[seqCounter].startIdx &&
+        instrCounter < sequences[seqCounter].endIdx) {
+      return getModule()->getFunction(sequences[seqCounter].func)->body
+               ? InSkipSeq
+               : InSeq;
+    }
+    return NotInSeq;
+  }
+  void maybeBeginSeq() {
+    instrCounter++;
+    auto currState = getCurrState();
+    if (currState != state) {
+      switch (currState) {
+        case NotInSeq:
+          return;
+        case InSeq:
+          transitionToInSeq();
+          break;
+        case InSkipSeq:
+          transitionToInSkipSeq();
+          break;
+      }
+    }
+    state = currState;
+  }
+  void transitionToInSeq() {
+    Function* outlinedFunc =
+      getModule()->getFunction(sequences[seqCounter].func);
+    ASSERT_ERR(outlinedBuilder.visitFunctionStart(outlinedFunc));
+
+    // Add a local.get instruction for every parameter of the outlined function.
+    Signature sig = outlinedFunc->type.getSignature();
+    for (Index i = 0; i < sig.params.size(); i++) {
+      ASSERT_ERR(outlinedBuilder.makeLocalGet(i));
+    }
+
+    // Make a call from the existing function to the outlined function. This
+    // call will replace the instructions moved to the outlined function
+    ASSERT_ERR(existingBuilder.makeCall(outlinedFunc->name, false));
+    DBG(std::cout << "\ncreated outlined fn: " << outlinedFunc->name);
+  }
+
+  void transitionToInSkipSeq() {
+    Function* outlinedFunc =
+      getModule()->getFunction(sequences[seqCounter].func);
+    ASSERT_ERR(existingBuilder.makeCall(outlinedFunc->name, false));
+    DBG(std::cout << "\n\nstarting to skip instructions "
+                  << sequences[seqCounter].startIdx << " - "
+                  << sequences[seqCounter].endIdx - 1 << " to "
+                  << sequences[seqCounter].func
+                  << " and adding call() instead");
+  }
+  void maybeEndSeq() {
+    if (instrCounter + 1 == sequences[seqCounter].endIdx) {
```

Comment:
Surprised this `+ 1` is still necessary since we've changed the boundary check to use `<=`.

Reply:
I think you're referring to the "<=" we changed to a "<" in getCurrState? When we need to end a sequence, we'll be on an index that is one less than the sequence endIdx. For example, in the first test in outlining.wast, the last instruction to be outlined is index 9, and the sequence endIdx is 10. We need to wrap up the outlined function at index 9 in case there is no next instruction.

Reply:
Makes sense, thanks.